File: LC-Java/src/Tree/_107_levelOrderBottom.java
package Tree;

import DataStructure.TreeNode;

import java.util.*;

@SuppressWarnings({"all"})
public class _107_levelOrderBottom {
    public List<List<Integer>> levelOrderBottom(TreeNode root) {
        List<List<Integer>> res = new ArrayList<>();
        if (root == null) {
            return res;
        }
        Queue<TreeNode> queue = new LinkedList<>();
        TreeNode cur = root;
        queue.add(cur);
        while (!queue.isEmpty()) {
            int size = queue.size();
            List<Integer> list = new ArrayList<>();
            while (size-- > 0) {
                TreeNode node = queue.poll();
                if (node == null) {
                    continue;
                }
                list.add(node.val);
                queue.add(node.left);
                queue.add(node.right);
            }
            if (!list.isEmpty()) {
                res.add(list);
            }
        }
        if (!res.isEmpty()) {
            Collections.reverse(res);
        }

        return res;

    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left=new TreeNode(2);
        root.right=new TreeNode(3);
        root.left.left=new TreeNode(4);
        root.left.right=new TreeNode(5);
        root.right.right=new TreeNode(7);
        root.right.left=new TreeNode(6);
        root.left.left.left=new TreeNode(8);
        root.left.left.right=new TreeNode(9);

        List<List<Integer>> res = new _107_levelOrderBottom().levelOrderBottom(root);
        System.out.println(res);
    }
}

File: TopKFromLCTags/Trees/107-levelOrderBottom.go
package trees

func levelOrderBottom(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	res := levelOrder107(root)
	for i := 0; i < len(res)>>1; i++ {
		res[i], res[len(res)-i-1] = res[len(res)-i-1], res[i]
	}

	return res
}

func levelOrder107(root *TreeNode) [][]int {
	if root == nil {
		return [][]int{}
	}
	queue, res, cur := make([]*TreeNode, 0), [][]int{}, root
	queue = append(queue, cur)
	for len(queue) > 0 {
		size := len(queue)
		tmp := make([]int, 0)
		for size > 0 {
			size--
			node := queue[0]
			queue = queue[1:]
			tmp = append(tmp, node.Val)
			if node.Left != nil {
				queue = append(queue, node.Left)
			}
			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}
		res = append(res, tmp)
	}
	return res
}
